src/compiler/checker.ts

```diff
@@ -6645,6 +6642,41 @@ module ts {
                     callIsIncomplete = !!templateLiteral.isUnterminated;
                 }
             }
+            else if (node.kind === SyntaxKind.Decorator) {
+                let decorator = <Decorator>node;
+                switch (decorator.parent.kind) {
+                    case SyntaxKind.ClassDeclaration:
+                    case SyntaxKind.ClassExpression:
+                        // A class decorator will have one argument (see `ClassDecorator` in core.d.ts)
+                        adjustedArgCount = 1;
+                        typeArguments = undefined;
+                        break;
+
+                    case SyntaxKind.PropertyDeclaration:
+                        // A property declaration decorator will have two arguments (see 
+                        // `PropertyDecorator` in core.d.ts)
+                        adjustedArgCount = 2;
+                        typeArguments = undefined;
+                        break;
+
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.GetAccessor:
+                    case SyntaxKind.SetAccessor:
+                        // A method or accessor declaration decorator will have two or three arguments (see 
+                        // `MethodDecorator` in core.d.ts)
+                        adjustedArgCount = signature.parameters.length >= 3 ? 3 : 2;

```

Comment:
Yes, but the way it worked when using an assignability check, you could have used a function with only two arguments for a method which I want to continue to allow for an "annotation"-style decorator.